src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,281 @@
 
-/* Handle and complete a matched XPMEM LMT receive request. Input parameters
- * include send buffer info (see definition in MPIDI_SHM_ctrl_xpmem_send_lmt_req_t)
- * and receive request. */
-MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_recv(uint64_t src_offset, uint64_t src_data_sz,
-                                                         uint64_t sreq_ptr, int src_lrank,
-                                                         MPIR_Request * rreq)
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE
+      category    : CH4
+      type        : int
+      default     : 32768
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        Chunk size is used in XPMEM COOP P2P and determines the concurrency of
+        copy and overhead of atomic per chunk copy. Best chunk size should
+        balance these two factors.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_do_lmt_coop_copy(const void *src_buf,
+                                                          size_t data_sz, void *dest_buf,
+                                                          OPA_int_t * counter_ptr,
+                                                          int ack_type,
+                                                          uint64_t req_ptr,
+                                                          MPIR_Request * local_req,
+                                                          int (*xpmem_lmt_fin_cb)
+                                                           (MPIDI_SHM_ctrl_hdr_t *))
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t copy_sz;
+    int cur_chunk = 0, total_chunk = 0;
+    uint64_t cur_offset;
+
+    MPIDI_SHM_ctrl_hdr_t ack_ctrl_hdr;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_send_fin_t *lmt_fin_hdr = &ack_ctrl_hdr.xpmem_slmt_send_fin;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_DO_LMT_COOP_COPY);
+
+    total_chunk =
+        data_sz / MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE +
+        (data_sz % MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE == 0 ? 0 : 1);
+
+    /* TODO: implement OPA_fetch_and_incr_int uint64_t type */
+    while ((cur_chunk = OPA_fetch_and_incr_int(counter_ptr)) < total_chunk) {
+        cur_offset = ((uint64_t) cur_chunk) * MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE;
+        copy_sz =
+            cur_offset + MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE <=
+            data_sz ? MPIR_CVAR_CH4_XPMEM_COOP_COPY_CHUNK_SIZE : data_sz - cur_offset;
+        mpi_errno =
+            MPIR_Localcopy(((char *) src_buf + cur_offset), copy_sz, MPI_BYTE,
+                           ((char *) dest_buf + cur_offset), copy_sz, MPI_BYTE);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    /* Send ack */
+    if (cur_chunk == total_chunk + 1) {
+        lmt_fin_hdr->req_ptr = req_ptr;
```

Comment:
Maybe I express my idea in a wrong way. The process that gets the `total_chunk + 1` chunk id should issue ack since it knows the copy has completed. 
if `cur_chunk == total_chunk-1`, it means copy is under processing
if `cur_chunk == total_chunk`, it means copy may finish

Reply:
I still do not get it. Can you please give me an example?

Reply:
1. sender gets cur_chunk == total_chunk - 1, it will copy last chunk
2. then receiver gets cur_chunk == total_chunk, it does not know whether copy is done and can only assume copy is being processed by sender, so receiver cannot complete this copy and has to wait `RECV_FIN` from sender
3. finally, sender gets cur_chunk == total_chunk + 1, sender knows copy is done and sends `RECV_FIN` to receiver

Reply:
I see. But what happens if only one process performs the copy? For example:
- receiver sent CTS to sender, and then performed all copies (last `cur_chunk == total_chunk`)
- sender is working on user compute and calls `MPI_Wait` only at the end of the program.

The actual data transfer has already been done, but the receive cannot complete until the sender finishes compute and makes an MPI call.

Reply:
that's right. as long as we perform cooperative copy, this waiting is unavoidable. It has nothing to do with this design.

Reply:
The caused delay can be very large, which results in even worse performance than single-copy. Let me think about if we can improve the synchronization. Can you please make two changes here:
1. Add comment to explain the design, even I was not able to understand it.
2. Add a `FIXME` comment to describe the issue of completion delay.